Add Feature_extraction.py for Batch & Parallel MS1/MS2 Feature Extraction

dev-coder2 opened the pull request:
Hi @singjc 
This pull request introduces a new script, Feature_extraction.py, which converts our existing feature extraction notebook into a standalone script. 

Key updates include:

- Batch and Parallel Processing: The script processes DIA-NN parquet files to extract MS1 and MS2 features in batches, utilizing GPU acceleration with cuDF for efficient data processing.

- Configuration Updates: Minor updates have been made to the dquartic_train_config.json to include the required file paths and settings for the pipeline. Currently, the MS1 and MS2 paths reflect the main outputs from DIA-NN. Once we generate the combined batch features for MS1 and MS2 through this script, these paths will be updated to point to the final CSV outputs.

- We can now extract MS1 and MS2 features by simply running:
"python Feature_extraction.py --config dquartic_train_config.json"

This enhancement streamlines our workflow and improves processing efficiency for the D4 diffusion model pipeline. Please review the changes and merge them into the main branch.

dev-coder2 committed:
add batch & parallel MS1/MS2 feature extraction script with config updates

singjc commented on dquartic_train_config.json:
Can you use a separate config file for the feature extraction, since a lot of the other fields in the training config is irrelevant for the feature extraction.

dev-coder2 replied:
ok i will make separate config

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
Can you put this file into a different module, either the `utils` module or create a new module (i.e. `data_preprocessing`). Can you also change the file name to lower case, `feature_extraction.py`

dev-coder2 replied:
ok i will update in new module

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+import cudf  # Ensure you have cuDF installed as part of RAPIDS
+from tqdm import tqdm
+
+# These imports should be adjusted based on your project structure
```
Why would these need to be adjusted based on the project structure?

dev-coder2 replied:
You're right to question that comment. I added it because I was initially uncertain about how the massdash package would be integrated into the project structure. I wasn't sure if users might need to adjust these import paths depending on their specific installation or if the package might be reorganized in the future. However, I now see this creates unnecessary confusion since massdash should be properly installed as a dependency just like cudf, pandas, etc.

I'll remove this comment in my updated code since it doesn't provide any useful information and could mislead other developers who might think there's something special about these imports. Thank you for pointing this out.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+    args = parser.parse_args()
+    
+    config = load_config(args.config)
+    data_config = config.get("data", {})
+    threads = config.get("threads", 4)
+    
+    # Map JSON configuration keys to processor parameters.
+    report_path = data_config.get("ms1_data_path", "report.parquet")
+    gen_lib_path = data_config.get("ms2_data_path", "gen_spec_lib.parquet")
+    mzml_path = data_config.get("raw_files", "/path/to/mzML/files/*.mzML")
+    
+    processor = DIAtoMassProcessor(
+        report_path=report_path,
+        gen_lib_path=gen_lib_path,
+        mzml_path=mzml_path,
+        batch_size=2000,
+        threads=threads,
+        output_dir="output_features",
+        rt_window=50,
+        mz_tol=20
+    )
+    
+    summary = processor.run_pipeline()
+    
+    # Print summary information and integration details
+    print("\nProcessing Summary:")
+    print(f"Total peptides processed: {summary['total_peptides']}")
+    print(f"MS1 features combined file: {summary['ms1_features_combined']}")
+    print(f"MS2 features combined file: {summary['ms2_features_combined']}")
+    
+    print("\nIntegration with diffusion-deconvolution-dia-msms-data:")
+    print("Add the following to your dquartic_train_config.json:")
+    print('{\n  "data": {\n'
+          f'    "ms1_data_path": "{summary["ms1_features_combined"]}",\n'
+          f'    "ms2_data_path": "{summary["ms2_features_combined"]}",\n'
+          '    "normalize": "minmax"\n'
+          '  }\n}')
+
+if __name__ == "__main__":
+    main()
```
Instead of having this as a separate python script, could we integrate it into the main dquartic cli?

dev-coder2 replied:
Hi @singjc
Yes, integration is possible, and I have already implemented it. You will see these changes once I commit after implementing the rest of your feedback.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+
+Features:
+    - Converts DIA-NN parquet files to MassDASH-compatible TSV format.
+    - Extracts MS1 and MS2 features in GPU-accelerated batch processing using cuDF.
```
I've never used cuDF before, but it seems pretty cool. Does it basically hold all of the dataframe data in the GPU? Will it still work if the user doesn't have a GPU? Will a default to dataframes being held on CPU?

dev-coder2 replied:
Hi @singjc ,
You're right about cuDF - it does hold dataframe data in GPU memory for accelerated processing. Currently, the code won't work if a user doesn't have a GPU since it directly creates cuDF DataFrames without checking for GPU availability. I'll modify the code to include a fallback mechanism that checks if cuDF is available and defaults to pandas (CPU-based) when no GPU is detected. This can be implemented using a try-except block that attempts to import cuDF, sets a GPU_AVAILABLE flag, and then conditionally uses either GPU or CPU processing paths based on this flag. This way, the code will be more robust and work on all systems.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+Features:
+    - Converts DIA-NN parquet files to MassDASH-compatible TSV format.
+    - Extracts MS1 and MS2 features in GPU-accelerated batch processing using cuDF.
+    - Exports per-batch CSV files and combines them into final output files.
```
Can you add some information on what the final output is? What are the columns etc? Also how large are the files?

dev-coder2 replied:
Okay, I will include these details.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+    mzml_path = data_config.get("raw_files", "/path/to/mzML/files/*.mzML")
+    
+    processor = DIAtoMassProcessor(
+        report_path=report_path,
+        gen_lib_path=gen_lib_path,
+        mzml_path=mzml_path,
+        batch_size=2000,
+        threads=threads,
+        output_dir="output_features",
+        rt_window=50,
+        mz_tol=20
```
Can you add the `batch_size`, `output_dir`, `rt_window` and the `mz_tol` to the config, so that the user can set these via the json config. Different datasets may have different m/z resolutions.

dev-coder2 replied:
Okay, I will include these in the new JSON file.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+        Returns:
+            tuple: (report_tsv_path, gen_lib_modified_tsv_path)
+        """
+        logging.info("Converting report.parquet to report.tsv...")
```
```suggestion
        logging.info(f"Converting {self.report_path} to report.tsv...")
```

dev-coder2 replied:
Changes implemented as suggested. The logging statement now uses the self.report_path variable.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+        report_tsv_path = os.path.join(os.path.dirname(self.output_dir), 'report.tsv')
+        report_df.to_csv(report_tsv_path, sep='\t', index=False)
+        
+        logging.info("Converting gen_spec_lib.parquet to gen_lib_modified.tsv...")
```
```suggestion
        logging.info(f"Converting {self.gen_lib_path} to gen_lib_modified.tsv...")
```

dev-coder2 replied:
Done

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+        plot_path = os.path.join(self.output_dir, "identifications_per_file.png")
+        plt.savefig(plot_path)
+        logging.info(f"Plot saved to {plot_path}")
+        return plot_path
```
Do we need to return the path?

dev-coder2 replied:
Yes, you are correct. We actually do not need the plot path, so I have removed the return statement.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+        self.plot_identifications(report_df)
+        
+        logging.info(f"Conversion complete. Files saved to {report_tsv_path} and {gen_lib_modified_tsv_path}")
+        return report_tsv_path, gen_lib_modified_tsv_path
```
Do we need to return the paths? Are they used else where after?

dev-coder2 replied:
You're right, these paths aren't used elsewhere, so I've removed the return statement.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+        logging.info(f"Loading peptides from {self.report_path}")
+        report_df = pd.read_parquet(self.report_path)
+        unique_peptides = report_df[['Modified.Sequence', 'Precursor.Charge']].drop_duplicates()
+        logging.info(f"Total unique peptides: {len(unique_peptides)}")
```
```suggestion
        logging.info(f"Total unique precursors: {len(unique_peptides)}")
```

Peptide is just the sequence i.e. `PEPTIDK`. Precursor is the ionized peptide sequence (i.e. has a charge state, `PEPTIDK_2`, `PEPTIDK_3`, `ASTYK_2` etc..)

dev-coder2 replied:
Thanks for improving my understanding. I've updated the logging statement as suggested.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+        
+        # Split unique peptides into batches.
+        peptide_batches = np.array_split(unique_peptides, int(np.ceil(len(unique_peptides) / self.batch_size)))
+        logging.info(f"Split peptides into {len(peptide_batches)} batches with batch size {self.batch_size}")
```
```suggestion
        logging.info(f"Split precursors into {len(peptide_batches)} batches with batch size {self.batch_size}")
```

dev-coder2 replied:
Done as above

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+        
+        # Process each batch sequentially
+        for batch_idx, batch in enumerate(peptide_batches):
+            logging.info(f"\nProcessing batch {batch_idx+1}/{len(peptide_batches)} with {len(batch)} peptides")
```
```suggestion
            logging.info(f"\nProcessing batch {batch_idx+1}/{len(peptide_batches)} with {len(batch)} precursors")
```

dev-coder2 replied:
Done as above

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+                    continue
+                
+                # Process each run's feature map
+                for run_name, feature_map in feature_map_collection.items():
```
Is there anyway to vectorize this/ remove the 3 level nested for loop? We also don't seem to be doing any parallelization. Could we parallelize the 3rd level for loop to process runs in parallel for the current peptide_charge?

dev-coder2 replied:
Yes, parallel processing of runs is possible. I've implemented this for the third loop based on your suggestion.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+        self.gen_lib_path = gen_lib_path
+        self.mzml_path = mzml_path
+        self.batch_size = batch_size
+        self.threads = threads
```
This doesn't seem to get used anywhere?

dev-coder2 replied:
You're right - the threads parameter wasn't being utilized properly in the original code. I've now implemented parallelization in the feature extraction process using joblib's Parallel/delayed functions, which makes effective use of the threads parameter to process multiple runs concurrently. The parameter now controls how many parallel workers are used when processing runs for each peptide/charge combination, significantly improving performance for multi-core systems.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+                    ms2_list.append(ms2_gpu.to_pandas())
+            
+            # Save current batch results as CSV files
+            if ms1_list:
+                batch_ms1_df = pd.concat(ms1_list, ignore_index=True)
+                batch_ms1_filename = os.path.join(self.ms1_dir, f"ms1_features_batch_{batch_idx}.csv")
+                batch_ms1_df.to_csv(batch_ms1_filename, index=False)
+                ms1_all_batches.append(batch_ms1_df)
+            if ms2_list:
+                batch_ms2_df = pd.concat(ms2_list, ignore_index=True)
+                batch_ms2_filename = os.path.join(self.ms2_dir, f"ms2_features_batch_{batch_idx}.csv")
+                batch_ms2_df.to_csv(batch_ms2_filename, index=False)
+                ms2_all_batches.append(batch_ms2_df)
```
Should we also ensure that there is matching data for both MS1 and MS2 and append only if that's the case? To avoid cases where there is ms1 extracted data, but no ms2 extracted data? Not sure if this actually happens or not.

dev-coder2 replied:
I believe this scenario is unlikely. Our data was initially processed with DIA-NN and subsequently re-extracted with MassDash. These tools are designed to ensure consistent and proper MS1 and MS2 data.

singjc commented on dquartic/dia-nn to massdash feature extraction/Feature_extraction.py:
```diff
@@ -0,0 +1,397 @@
+            logging.info(f"Finished processing batch {batch_idx+1} in {batch_elapsed:.2f} seconds.")
+        
+        # Combine all batch results and export final combined files.
+        if ms1_all_batches:
+            final_ms1_df = pd.concat(ms1_all_batches, ignore_index=True)
+            final_ms1_file = os.path.join(self.ms1_dir, "ms1_features_all_peptides_combined.csv")
+            final_ms1_df.to_csv(final_ms1_file, index=False)
+        else:
+            final_ms1_file = None
+        if ms2_all_batches:
+            final_ms2_df = pd.concat(ms2_all_batches, ignore_index=True)
+            final_ms2_file = os.path.join(self.ms2_dir, "ms2_features_all_peptides_combined.csv")
+            final_ms2_df.to_csv(final_ms2_file, index=False)
+        else:
+            final_ms2_file = None
```
If ms1_all_batches and ms2_all_batches are empty, should we throw an error or throw a logging error to say that there is no data being written? Which likely means something is wrong, or none of the peptides from the DIA-NN report were found with MassDash's extraction?

dev-coder2 replied:
Agreed, this is a good way to handle potential issues. I've modified the code to raise an error if empty files are encountered.

singjc reviewed (changes requested):
Great work! Thanks for implementing this. I added a few comments and suggestions.

dev-coder2 commented:
Hi @singjc,

I have taken your feedback, and I am looking at it point by point.
Just one request: can I send my GSoC proposal to you as a personal message on Discord for feedback?

singjc commented:
> Just one request: can I send my GSoC proposal to you as a personal message on Discord for feedback?

Sure, I will take a look

dev-coder2 commented:
> Sure, I will take a look

Hi @singjc, I've sent you my proposal. Could you please check it?

dev-coder2 commented:
Hi @singjc,

I have updated this PR to address all your feedback. Here's a summary of the changes:

**Structure and Integration**

- Renamed Feature_extraction.py to feature_extraction.py and moved it to the data_preprocessing module
- Created a dedicated feature_extraction_config.json file for better separation of concerns
- Integrated the feature extraction functionality into the main CLI with a new extract-features command

**GPU Acceleration and Fallback**

- Implemented cuDF's pandas accelerator mode that automatically falls back to standard pandas on systems without a GPU
- Added proper detection and logging of GPU availability

**Performance Optimization**

- Replaced the 3-level nested loop structure with joblib parallelization for the innermost loop
- Made effective use of the threads parameter to process runs concurrently
- Added proper vectorization of operations where possible

**Error Handling and Validation**

- Added explicit checks for empty feature lists with descriptive error messages
- Implemented validation to ensure data consistency between MS1 and MS2 results
- Improved error reporting throughout the pipeline

**Documentation and Code Cleanup**

- Updated the docstring with detailed information about output files, columns, and typical file sizes
- Removed unnecessary return statements from methods where return values weren't used
- Removed confusing comments about import adjustments
- Ensured consistent code style throughout

These improvements make the feature extraction process more robust, better integrated with the main project, and more efficient on both GPU and CPU systems while maintaining the same functionality.

dev-coder2 committed:
Improve feature extraction implementation
